File: packages/drivekit_driver_data/flutter_drivekit_driver_data_android/android/src/main/kotlin/com/drivequant/drivekit/flutter/driverdata/DriveKitDriverDataPlugin.kt
package com.drivequant.drivekit.flutter.driverdata

import android.content.Context
import com.drivequant.drivekit.databaseutils.entity.Route
import com.drivequant.drivekit.databaseutils.entity.Trip
import com.drivequant.drivekit.driverdata.DriveKitDriverData
import com.drivequant.drivekit.driverdata.trip.RouteQueryListener
import com.drivequant.drivekit.driverdata.trip.RouteStatus
import com.drivequant.drivekit.driverdata.trip.TripDeleteQueryListener
import com.drivequant.drivekit.driverdata.trip.TripQueryListener
import com.drivequant.drivekit.driverdata.trip.TripsQueryListener
import com.drivequant.drivekit.driverdata.trip.TripsSyncStatus
import com.drivequant.drivekit.flutter.driverdata.mapper.PigeonMapper
import io.flutter.embedding.engine.plugins.FlutterPlugin

class DriveKitDriverDataPlugin :
    FlutterPlugin,
    AndroidDriverDataApi {
    private var context: Context? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        AndroidDriverDataApi.setUp(flutterPluginBinding.binaryMessenger, this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        AndroidDriverDataApi.setUp(binding.binaryMessenger, null)
        context = null
    }

    override fun getTripsOrderByDateAsc(synchronizationType: PigeonSynchronizationType, transportationModes: List<PigeonTransportationMode>, callback: (Result<PigeonGetTripsResponse>) -> Unit) {
        DriveKitDriverData.getTripsOrderByDateAsc(
            type = PigeonMapper.fromPigeonSynchronizationType(synchronizationType),
            transportationModes = transportationModes.map { PigeonMapper.fromPigeonTransportationMode(it) },
            listener = object : TripsQueryListener {
                override fun onResponse(status: TripsSyncStatus, trips: List<Trip>) {
                    callback(
                        Result.success(
                            PigeonGetTripsResponse(
                                status = PigeonMapper.toPigeonTripsSyncStatus(status),
                                trips = trips.map { PigeonMapper.toPigeonTrip(it) }
                            )
                        )
                    )
                }
            }
        )
    }

    override fun getTripsOrderByDateDesc(synchronizationType: PigeonSynchronizationType, transportationModes: List<PigeonTransportationMode>, callback: (Result<PigeonGetTripsResponse>) -> Unit) {
        DriveKitDriverData.getTripsOrderByDateDesc(
            type = PigeonMapper.fromPigeonSynchronizationType(synchronizationType),
            transportationModes = transportationModes.map { PigeonMapper.fromPigeonTransportationMode(it) },
            listener = object : TripsQueryListener {
                override fun onResponse(status: TripsSyncStatus, trips: List<Trip>) {
                    callback(
                        Result.success(
                            PigeonGetTripsResponse(
                                status = PigeonMapper.toPigeonTripsSyncStatus(status),
                                trips = trips.map { PigeonMapper.toPigeonTrip(it) }
                            )
                        )
                    )
                }
            }
        )
    }

    override fun getTrip(itinId: String, callback: (Result<PigeonGetTripResponse>) -> Unit) {
        DriveKitDriverData.getTrip(
            itinId = itinId,
            listener = object : TripQueryListener {
                override fun onResponse(status: TripsSyncStatus, trip: Trip?) {
                    val pigeonTrip: PigeonTrip? = trip?.let {
                        PigeonMapper.toPigeonTrip(it)
                    }

                    callback(
                        Result.success(
                            PigeonGetTripResponse(
                                status = PigeonMapper.toPigeonTripsSyncStatus(status),
                                trip = pigeonTrip
                            )
                        )
                    )
                }
            }
        )
    }

    override fun getRoute(itinId: String, callback: (Result<PigeonGetRouteResponse>) -> Unit) {
        DriveKitDriverData.getRoute(
            itinId = itinId,
            listener = object : RouteQueryListener {
                override fun onResponse(status: RouteStatus, route: Route?) {
                    val pigeonRoute: PigeonRoute? = route?.let {
                        PigeonMapper.toPigeonRoute(it)
                    }
                    callback(
                        Result.success(
                            PigeonGetRouteResponse(
                                status = PigeonMapper.toRouteStatus(status),
                                route = pigeonRoute
                            )
                        )
                    )
                }
            }
        )
    }

    override fun deleteTrip(itinId: String, callback: (Result<Boolean>) -> Unit) {
        DriveKitDriverData.deleteTrip(
            itinId,
            object : TripDeleteQueryListener {
                override fun onResponse(status: Boolean) {
                    callback(Result.success(status))
                }
            }
        )
    }
}

File: packages/drivekit_driver_data/flutter_drivekit_driver_data_ios/ios/Classes/DriveKitDriverDataPlugin.swift
import Flutter
import UIKit
import DriveKitCoreModule
import DriveKitDriverDataModule
import DriveKitDBTripAccessModule

extension FlutterError: Error {}

public class DriveKitDriverDataPlugin: NSObject, FlutterPlugin, IOSDriverDataApi {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger: FlutterBinaryMessenger = registrar.messenger()
        let api: IOSDriverDataApi & NSObjectProtocol = DriveKitDriverDataPlugin.init()

        IOSDriverDataApiSetup.setUp(binaryMessenger: messenger, api: api)
    }

    func getTripsOrderByDateAsc(synchronizationType: PigeonSynchronizationType, transportationModes: [PigeonTransportationMode], completion: @escaping (Result<PigeonGetTripsResponse, any Error>) -> Void) {
        DriveKitDriverData.shared.getTripsOrderByDateAsc(
            withTransportationModes: transportationModes.map({ TransportationMode(from: $0) }),
            type: SynchronizationType(from: synchronizationType)
        ) { status, trips in
            completion(Result.success(PigeonGetTripsResponse(from: status, trips: trips)))
        }
    }

    func getTripsOrderByDateDesc(synchronizationType: PigeonSynchronizationType, transportationModes: [PigeonTransportationMode], completion: @escaping (Result<PigeonGetTripsResponse, any Error>) -> Void) {
        DriveKitDriverData.shared.getTripsOrderByDateDesc(
            withTransportationModes: transportationModes.map({ TransportationMode(from: $0) }),
            type: SynchronizationType(from: synchronizationType)
        ) { status, trips in
            completion(Result.success(PigeonGetTripsResponse(from: status, trips: trips)))
        }
    }

    func getTrip(itinId: String, completion: @escaping (Result<PigeonGetTripResponse, any Error>) -> Void) {
        DriveKitDriverData.shared.getTrip(
            itinId: itinId
        ) { status, trip  in
            completion(Result.success(PigeonGetTripResponse(from: status, trip: trip)))
        }
    }

    func getRoute(itinId: String, completion: @escaping (Result<PigeonGetRouteResponse, any Error>) -> Void) {
        DriveKitDriverData.shared.getRoute(itinId: itinId) { route in
            let pigeonRouteResponse = PigeonGetRouteResponse(
                from: route)
            completion(Result.success(pigeonRouteResponse))
        }
    }

    func deleteTrip(itinId: String, completion: @escaping (Result<Bool, any Error>) -> Void) {
        DriveKitDriverData.shared.deleteTrip(itinId: itinId) { status in
            completion(Result.success(status))
        }
    }
    
    func updateDriverPassengerMode(itinId: String, mode: PigeonDriverPassengerMode, comment: String, completion: @escaping (Result<PigeonUpdateDriverPassengerModeStatus, any Error>) -> Void) {
        DriveKitDriverData.shared.updateDriverPassengerMode(itinId: itinId, mode: DKDriverPassengerMode.passenger, comment: comment) {
            status in
            completion(Result.success(PigeonUpdateDriverPassengerModeStatus(from: status)))
        }
    }
}
